JSR_Shopping_List_Functions-master/js/app.js

```diff
@@ -0,0 +1,947 @@
+// 			chai.assert.throws(() => {
+// 				removeNthItem(1, [])
+// 			}, Error);
+// 		});
+
+// 		it('should throw error if i is not a number', () => {
+
+// 			// if i is not a number
+// 			chai.assert.throws(() => {
+// 				removeNthItem('adfas', [])
+// 			}, Error);
+// 		})
+// 	});
+
+	/* 6
+		@function removeNItems
+		@param i {number}
+		@param num {number}
+		@param list {array, []}
+		@returns list
+		@description
+			same as above but now we wish to remove ALL
+			items from i to i+num and return the resulting list
+			if...
+				- `i` < 0
+				- `i` or `num` is not a number
+				- `i+num` > length of list
+				- `num` > length of list
+			^^ return error
+	*/
+
+	// // implement function here
+	// 	const removeNItems = (i, num, list) =>{
+	// 			if(i < 0 || typeof (i + num) === 'string' || num > list.length || (i+num) > list.length ){
+	// 			throw new error('bad');
+	// 			}
+	// 			else{
+
+	// 			list.splice(i, num);
+	// 			return list;
+	// 			}
+				
+
+	// 	}
+
+
+		
+	// // TEST
+	// describe('6. removeNItems', () => {
+	// 	it('should remove i-th item from list', () => {
+	// 		let list = addToShoppingList({
+	// 			'item': 'test',
+	// 			'price': 1
+	// 		});
+	// 		list = addToShoppingList({
+	// 			'item': 'test2',
+	// 			'price': 2
+	// 		}, list);
+	// 		list = addToShoppingList({
+	// 			'item': 'test3',
+	// 			'price': 3
+	// 		}, list);
+
+			
+	// 		list = removeNItems(1, 1, list);
+	// 		console.log(list);
+	// 		chai.assert.equal(list.length, 1)
+
+	// 		chai.assert.equal(list[0].item, 'test')
+	// 		chai.assert.equal(list[0].price, 1)
+	// 	});
+
+	// 	it('should throw error if i + num < 0', () => {
+	// 		// if i < 0
+	// 		chai.assert.throws(() => {
+	// 			removeNItems(-1, 0, [])
+	// 		}, Error);
+	// 	});
+
+	// 	it('should throw error if i + num > length of list', () => {
+	// 		// if i > length of array
+	// 		chai.assert.throws(() => {
+	// 			removeNItems(1, 2, ['a', 'b'])
+	// 		}, Error);
+	// 	});
+
+	// 	it('should throw error if i is not a number', () => {
+
+	// 		// if i is not a number
+	// 		chai.assert.throws(() => {
+	// 			removeNItems('adfas', 1, [])
+	// 		}, Error);
+	// 	})
+
+	// 	it('should throw error if num is not a number', () => {
+
+	// 		// if i is not a number
+	// 		chai.assert.throws(() => {
+	// 			removeNItems(1, 'asasdfa', [])
+	// 		}, Error);
+	// 	})
+
+	// 	it('should throw error if num > length of list', () => {
+
+	// 		// if i is not a number
+	// 		chai.assert.throws(() => {
+	// 			removeNItems(1, 8, [])
+	// 		}, Error);
+	// 	})
+	// });
+
+	// /* 7
+	// 	@function smartRemoveItems
+	// 	@param i {number}
+	// 	@param list {array, []}
+	// 	@returns list
+	// 	@description
+	// 		- if `i` is < 0, remove i number of items
+	// 			from END of list
+	// 		- if `i` > length of list, return list immediately
+	// 		- if `i` > 0 remove i number of items
+	// 			from START of list
+	// */
+
+	// // implement function here
+	// const smartRemoveItems = (i, list,) =>{
```

Comment:
This one always tests my memory of whether to use `splice` or `slice`. The difference here is that `splice` returns the _deleted_ items, whereas `slice` returns a shallow copy of the altered array _without actually altering the original_, so you have to return from the `slice` operation _or_ assign a new variable and return that one.

This is a bit whack of a test because though your array lengths were right for third case, the actual values would be wrong because it's returning the deleted elements instead of the ones we wanted to keep.

[Taq's solution](https://github.com/FEWDMaterials/JSR_Shopping_List_Functions/blob/solutions/JSR0312/js/app.js#L330) is incredibly terse and I probably wouldn't go this cryptic in it, but consider this:

```
const smartRemoveItems = (i, list) =>{
	if (i < 0) {
		return list.slice(0, i);
	}
	else if (i > list.length){
		return list;
	}
	else {
		return list.slice(i);
	}
}
```